source/common/protobuf/utility.cc

```diff
@@ -852,6 +907,31 @@ ProtobufWkt::Value ValueUtil::listValue(const std::vector<ProtobufWkt::Value>& v
   return val;
 }
 
+void ValueUtil::appendJsonToString(const ProtobufWkt::Value& value, std::string& dest) {
+  JsonStringStreamer streamer(dest);
+  switch (value.kind_case()) {
+  case ProtobufWkt::Value::KIND_NOT_SET:
+  case ProtobufWkt::Value::kNullValue:
+    streamer.addNull();
+    break;
+  case ProtobufWkt::Value::kNumberValue:
+    streamer.addNumber(value.number_value());
+    break;
+  case ProtobufWkt::Value::kStringValue:
+    streamer.addString(value.string_value());
+    break;
+  case ProtobufWkt::Value::kBoolValue:
+    streamer.addBool(value.bool_value());
+    break;
+  case ProtobufWkt::Value::kStructValue:
+    structValueToJson(value.struct_value(), *streamer.makeRootMap());
+    break;
+  case ProtobufWkt::Value::kListValue:
+    listValueToJson(value.list_value(), *streamer.makeRootArray());
```

Comment:
I'm not seeing how renaming makeRootArray to addArray in the streamer here would eliminate duplication; can you explain?

Reply:
If we rename the makeRoot* to add*, then we can update our two function to following to eliminate the duplicated code:

```
template <class StreamerOrLevel>
void valueToJson(const ProtobufWkt::Value& value, StreamerOrLevel& streamer_or_level) {
  switch (value.kind_case()) {
  case ProtobufWkt::Value::KIND_NOT_SET:
  case ProtobufWkt::Value::kNullValue:
    streamer_or_level.addNull();
    break;
  case ProtobufWkt::Value::kNumberValue:
    streamer_or_level.addNumber(value.number_value());
    break;
  case ProtobufWkt::Value::kStringValue:
    streamer_or_level.addString(value.string_value());
    break;
  case ProtobufWkt::Value::kBoolValue:
    streamer_or_level.addBool(value.bool_value());
    break;
  case ProtobufWkt::Value::kStructValue:
    structValueToJson(value.struct_value(), *streamer_or_level.addMap());
    break;
  case ProtobufWkt::Value::kListValue:
    listValueToJson(value.list_value(), *streamer_or_level.addArray());
    break;
  }
}
```

```
void ValueUtil::appendJsonToString(const ProtobufWkt::Value& value, std::string& dest) {
  JsonStringStreamer streamer(dest);
  valueToJson(value, streamer);
}
```

Reply:
You can compare this with https://github.com/envoyproxy/envoy/pull/36334#discussion_r1777180288

Reply:
I see, OK. I think I am OK with the way you have it in this PR with a duplicated switch. I can see the benefit of typing less code, but I feel like it creates an external dependency that Json::StreamerBase and Json::StreamerBase::Level have to have the same API, and if we are going to do that I think they should probably implement the same interface. I think it's probably not worth it to save this duplicated switch.

I am really a big fan of C++in most ways but one that would be nice is a language feature for describing an interface that a class passed into a template must implement.

Reply:
Concept? Haha.